### PR TITLE
[Feature] support concurrency limit for resource group

### PR DESCRIPTION
**Pull request by murphyatwork**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #

### CODE REVIEW

**Comment by wangruin** [comment]
run starrocks_tscancode

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 11 / 11 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/WorkGroup.java|10|10|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/WorkGroupAnalyzer.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 11 / 11 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/WorkGroup.java|10|10|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/WorkGroupAnalyzer.java|1|1|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/WorkGroupAnalyzer.java [hunk 1]
```diff
@@ -153,7 +153,7 @@ public static void analyzeProperties(WorkGroup workgroup, Map<String, String> pr
 
             if (key.equalsIgnoreCase(WorkGroup.CONCURRENCY_LIMIT)) {
                 int concurrencyLimit = Integer.parseInt(value);
-                if (concurrencyLimit <= 0) {
+                if (concurrencyLimit < 0) {
```
Comment (silverbullet233) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/WorkGroupAnalyzer.java:156]: why remove equal ?

Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/WorkGroupAnalyzer.java:156]: `concurrenctLimit == 0` should means disable this limitation.

**Review by silverbullet233 — DISMISSED** [review]

File: be/src/exec/pipeline/fragment_executor.cpp [hunk 2]
```diff
@@ -141,23 +140,23 @@ Status FragmentExecutor::_prepare_workgroup(const TExecPlanFragmentParams& reque
             wg = WorkGroupManager::instance()->get_default_workgroup();
         }
         DCHECK(wg != nullptr);
-        if (_query_ctx->init_query(wg.get())) {
-            _wg = wg;
-        }
+        RETURN_IF_ERROR(_query_ctx->init_query(wg.get()));
+        _wg = wg;
```
Comment (stdpain) [inline be/src/exec/pipeline/fragment_executor.cpp:144]: if a query_ctx created success and it has two fragment_ctx.
In this case if fragment 1 and fragment 2 both failed, _wg->decr_num_queries() will execute twice.

Comment (mofeiatwork) [reply be/src/exec/pipeline/fragment_executor.cpp:144]: The `_wg->decr_num_queries()`will only be executed after `QueryContextManager::remove`

File: be/src/exec/pipeline/pipeline_driver.cpp [hunk 3]
```diff
@@ -317,12 +317,20 @@ void PipelineDriver::finalize(RuntimeState* runtime_state, DriverState state) {
...
             DCHECK(!this->is_still_pending_finish());
-            if (_fragment_ctx->enable_resource_group()) {
-                if (_workgroup) {
+
+            auto frag_id = _fragment_ctx->fragment_instance_id();
+            int active_fragments = _query_ctx->num_active_fragments();
+            int active_drivers = _fragment_ctx->num_drivers();
+            int64_t wg_id = _workgroup != nullptr ? _workgroup->id() : 0;
+            if (ExecEnv::GetInstance()->query_context_mgr()->remove(query_id)) {
+                if (wg_id != 0) {
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/pipeline_driver.cpp:326]: `0` is `WorkGroup::DEFAULT_WG_ID`, representing the default workgroup, so `_workgroup->id()` may return `0`. Maybe we could just check `_workgroup != nullptr` here.

File: be/src/exec/pipeline/pipeline_driver_executor.cpp [hunk 4]
```diff
@@ -114,21 +114,19 @@ void GlobalDriverExecutor::_worker_thread() {
...
-                if (is_big_query) {
+            if (!status.ok() || !bigquery_status.ok()) {
+                if (!bigquery_status.ok()) {
                     LOG(WARNING) << "[Driver] Process exceed limit, query_id="
                                  << print_id(driver->query_ctx()->query_id())
-                                 << ", instance_id=" << print_id(driver->fragment_ctx()->fragment_instance_id());
-                    query_ctx->cancel(Status::Cancelled("exceed big query limit"));
+                                 << ", instance_id=" << print_id(driver->fragment_ctx()->fragment_instance_id())
+                                 << ", status=" << bigquery_status;
+                    query_ctx->cancel(bigquery_status);
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/pipeline_driver_executor.cpp:129]: Could we simplify this code by merging `status` and `bigquery_status` like follows?
```C++
auto status = driver->process(runtime_state, worker_id);
this->_driver_queue->update_statistics(driver);

if (status.ok() && driver->workgroup() != nullptr) {
    // Check big query
    status = driver->workgroup()->check_big_query(*query_ctx);
}


if (!status.ok()) {
    LOG(WARNING) << "[Driver] Process error, query_id="
                    << print_id(driver->query_ctx()->query_id())
                    << ", instance_id=" << print_id(driver->fragment_ctx()->fragment_instance_id())
                    << ", status=" << status;
    query_ctx->cancel(status);
    // ...
}
```

**Review by ZiheLiu — DISMISSED** [review]

**Review by stdpain — APPROVED** [review]

File: be/src/exec/pipeline/pipeline_driver.cpp (hunk 3 above)
Comment (mofeiatwork) [reply be/src/exec/pipeline/pipeline_driver.cpp:326]: done

**Review by stdpain — DISMISSED** [review]

File: be/src/exec/pipeline/pipeline_driver_executor.cpp (hunk 4 above)
Comment (mofeiatwork) [reply be/src/exec/pipeline/pipeline_driver_executor.cpp]: done

File: be/src/exec/pipeline/pipeline_driver_executor.cpp [hunk 5]
```diff
@@ -111,30 +111,20 @@ void GlobalDriverExecutor::_worker_thread() {
                 continue;
             }
 
-            auto status = driver->process(runtime_state, worker_id);
+            auto maybeState = driver->process(runtime_state, worker_id);
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/pipeline_driver_executor.cpp:114]: `maybe_state`?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 11 / 11 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/WorkGroup.java|10|10|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/WorkGroupAnalyzer.java|1|1|100.00%|

**Review by ZiheLiu — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 11 / 11 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/WorkGroup.java|10|10|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/WorkGroupAnalyzer.java|1|1|100.00%|

File: be/src/exec/pipeline/pipeline_driver_executor.cpp (hunk 5 above)
Comment (mofeiatwork) [reply be/src/exec/pipeline/pipeline_driver_executor.cpp]: done

**Review by silverbullet233 — APPROVED** [review]

**Review by stdpain — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**